In [1]:
from pathlib import Path
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.colors as colors
from IPython.display import clear_output
from matplotlib import cm
from scipy.optimize import minimize
from scipy.stats import norm, qmc
from sklearn.gaussian_process import GaussianProcessRegressor
from sklearn.gaussian_process.kernels import Matern, WhiteKernel, ConstantKernel as C
from sklearn.ensemble import RandomForestRegressor

In [2]:
# Load all data
folder = Path(r'C:\Users\Dave Amin\Documents\docs-Sync\Career\2025-10 IC AI course\Modules\Capstone2\initial_data')
folder2 = Path(r'function_8')

def load_data():
    file_path_inputs =  folder / folder2 / rf"initial_inputs.npy"
    file_path_outputs = folder / folder2 / rf"initial_outputs.npy"
    X = np.load(file_path_inputs)
    Y = np.load(file_path_outputs)
    print('Initial data :','X.shape', X.shape, 'Y.shape', Y.shape)
    
    X_new_point = np.array([0.00000e+00, 0.00000e+00, 8.57000e-04, 0.00000e+00, 1.00000e+00, 1.15384e-01, 0.00000e+00, 4.16099e-01], dtype=np.float64)
    Y_new_point = np.array([9.6336548314169], dtype=np.float64)
    X = np.vstack((X, X_new_point))
    Y = np.append(Y, Y_new_point)
    print('Added week 1 :','X.shape', X.shape, 'Y.shape', Y.shape)

    X_new_point = np.array([0.191681, 0.218924, 0.113081, 0.212720, 1.000000, 0.770107, 0.277017, 1.000000], dtype=np.float64)
    Y_new_point = np.array([9.852825086592], dtype=np.float64)
    X = np.vstack((X, X_new_point))
    Y = np.append(Y, Y_new_point)
    print('Added week 2 :','X.shape', X.shape, 'Y.shape', Y.shape)

    X_new_point = np.array([0.087719, 0.102828, 0.213350, 0.000000, 1.000000, 0.512206, 0.213228, 0.015300], dtype=np.float64)
    Y_new_point = np.array([9.89944513359], dtype=np.float64)
    X = np.vstack((X, X_new_point))
    Y = np.append(Y, Y_new_point)
    print('Added week 3 :','X.shape', X.shape, 'Y.shape', Y.shape)

    X_new_point = np.array([0.000000, 0.000000, 0.000000, 0.845450, 1.000000, 0.811161, 0.000000, 0.000000], dtype=np.float64)
    Y_new_point = np.array([9.190328129579], dtype=np.float64)
    X = np.vstack((X, X_new_point))
    Y = np.append(Y, Y_new_point)
    print('Added week 4 :','X.shape', X.shape, 'Y.shape', Y.shape)

    X_new_point = np.array([0.000000, 0.000000, 0.000000, 0.000000, 1.000000, 0.508290, 0.000000, 0.079755], dtype=np.float64)
    Y_new_point = np.array([9.7571657898975], dtype=np.float64)
    X = np.vstack((X, X_new_point))
    Y = np.append(Y, Y_new_point)
    print('Added week 5 :','X.shape', X.shape, 'Y.shape', Y.shape)

    X_new_point = np.array([0.082682, 0.917557, 0.413349, 0.038587, 0.180483, 0.597658, 0.091566, 0.094504], dtype=np.float64)
    Y_new_point = np.array([8.906477374608901], dtype=np.float64)
    X = np.vstack((X, X_new_point))
    Y = np.append(Y, Y_new_point)
    print('Added week 6 :','X.shape', X.shape, 'Y.shape', Y.shape)

    X_new_point = np.array([0.077719, 0.102828, 0.213350, 0.000000, 1.000000, 0.512206, 0.223228, 0.015300], dtype=np.float64)
    Y_new_point = np.array([9.89802477359], dtype=np.float64)
    X = np.vstack((X, X_new_point))
    Y = np.append(Y, Y_new_point)
    print('Added week 7 :','X.shape', X.shape, 'Y.shape', Y.shape)

    X_new_point = np.array([0.092719, 0.102828, 0.213350, 0.000000, 1.000000, 0.512206, 0.213228, 0.015300], dtype=np.float64)
    Y_new_point = np.array([9.89964075359], dtype=np.float64)
    X = np.vstack((X, X_new_point))
    Y = np.append(Y, Y_new_point)
    print('Added week 8 :','X.shape', X.shape, 'Y.shape', Y.shape)

    X_new_point = np.array([0.102719, 0.102828, 0.213350, 0.000000, 1.000000, 0.512206, 0.210228, 0.015300], dtype=np.float64)
    Y_new_point = np.array([9.89987272959], dtype=np.float64)
    X = np.vstack((X, X_new_point))
    Y = np.append(Y, Y_new_point)
    print('Added week 9 :','X.shape', X.shape, 'Y.shape', Y.shape)

    X_new_point = np.array([0.107719, 0.102828, 0.213350, 0.000000, 1.000000, 0.512206, 0.208728, 0.015300], dtype=np.float64)
    Y_new_point = np.array([9.89982521759], dtype=np.float64)
    X = np.vstack((X, X_new_point))
    Y = np.append(Y, Y_new_point)
    print('Added week 10 :','X.shape', X.shape, 'Y.shape', Y.shape)

    # next week new values go here !!!!
  
    return X,Y
X,Y = load_data()
n_dims = X.shape[1]

Initial data : X.shape (40, 8) Y.shape (40,)
Added week 1 : X.shape (41, 8) Y.shape (41,)
Added week 2 : X.shape (42, 8) Y.shape (42,)
Added week 3 : X.shape (43, 8) Y.shape (43,)
Added week 4 : X.shape (44, 8) Y.shape (44,)
Added week 5 : X.shape (45, 8) Y.shape (45,)
Added week 6 : X.shape (46, 8) Y.shape (46,)
Added week 7 : X.shape (47, 8) Y.shape (47,)
Added week 8 : X.shape (48, 8) Y.shape (48,)
Added week 9 : X.shape (49, 8) Y.shape (49,)
Added week 10 : X.shape (50, 8) Y.shape (50,)


In [3]:
# print('X.shape', X.shape)
# print('X col1 min',np.min(X[:, 0]), 'max',np.max(X[:, 0]))
# print('X col2 min',np.min(X[:, 1]), 'max',np.max(X[:, 1]))
# print('Y.shape', Y.shape)
print("Min Y =",min(Y),"Max Y =",max(Y))

Min Y = 5.5921933895401965 Max Y = 9.89987272959


In [4]:
print("X",X)
print("Y",Y)

X [[6.04994453e-01 2.92215020e-01 9.08452748e-01 3.55506242e-01
  2.01668719e-01 5.75338005e-01 3.10310951e-01 7.34281377e-01]
 [1.78006959e-01 5.66222654e-01 9.94861845e-01 2.10325006e-01
  3.20152657e-01 7.07908792e-01 6.35384489e-01 1.07131627e-01]
 [9.07697668e-03 8.11626153e-01 5.20520360e-01 7.56866752e-02
  2.65111825e-01 9.16516894e-02 5.92415145e-01 3.67320262e-01]
 [5.06028164e-01 6.53730123e-01 3.63410779e-01 1.77981049e-01
  9.37283044e-02 1.97425331e-01 7.55826900e-01 2.92472339e-01]
 [3.59909264e-01 2.49075679e-01 4.95997170e-01 7.09214981e-01
  1.14987195e-01 2.89206921e-01 5.57295151e-01 5.93881726e-01]
 [7.78818344e-01 3.41949948e-03 3.37983130e-01 5.19527778e-01
  8.20906993e-01 5.37246689e-01 5.51347098e-01 6.60032086e-01]
 [9.08649322e-01 6.22496998e-02 2.38259546e-01 7.66603545e-01
  1.32335962e-01 9.90243814e-01 6.88067822e-01 7.42495941e-01]
 [5.86371444e-01 8.80735726e-01 7.45020752e-01 5.46034849e-01
  9.64887799e-03 7.48991763e-01 2.30907070e-01 9.79156228e-02

## 10 April - v1 - The "Precision Pin"

In [5]:
import numpy as np

X_all = np.array(X)
Y_all = np.array(Y)
best_idx = np.argmax(Y_all)
x_best = X_all[best_idx]

# Momentum Factor: How much further to push the current winning vector
# We reduce the step size to 0.005 to avoid overshooting the peak
step = 0.005 

# Candidate C: "Steady Vector"
# Continue the exact direction that gave us the 9.8998 record
cand_c = x_best.copy()
cand_c[0] = np.clip(cand_c[0] + step, 0, 1)
cand_c[6] = np.clip(cand_c[6] - (step * 0.3), 0, 1) # Proportional decrease

# Candidate D: "Third-Variable Synergy" (The Wildcard)
# In many models, as LR (X0) goes up, a tiny nudge in regularization (X1) helps
cand_d = cand_c.copy()
cand_d[1] = np.clip(cand_d[1] + 0.002, 0, 1) 

print(f"Current Best: {Y_all[best_idx]:.6f}")
print(f"Cand C (Steady): {'-'.join(f'{v:.6f}' for v in cand_c)}")
print(f"Cand D (Synergy): {'-'.join(f'{v:.6f}' for v in cand_d)}")

Current Best: 9.899873
Cand C (Steady): 0.107719-0.102828-0.213350-0.000000-1.000000-0.512206-0.208728-0.015300
Cand D (Synergy): 0.107719-0.104828-0.213350-0.000000-1.000000-0.512206-0.208728-0.015300


## 2 April - v1 - Nudge

In [5]:
import numpy as np

# Updated Data with your new 9.8996 point
X_all = np.array(X)
Y_all = np.array(Y)
best_idx = np.argmax(Y_all)
x_best = X_all[best_idx]

# STRATEGY: "Momentum + Balance"
# We double down on Dim 0 and test a micro-nudge on Dim 6
step_momentum = 0.010  # Increasing the step size since we found a lead
step_micro = 0.003     # Tiny nudge for secondary variables

# Candidate A: Pure Momentum (Follow the Dim 0 slope)
cand_a = x_best.copy()
cand_a[0] = np.clip(cand_a[0] + step_momentum, 0, 1)

# Candidate B: The Pivot (Dim 0 Momentum + Dim 6 Nudge)
cand_b = x_best.copy()
cand_b[0] = np.clip(cand_b[0] + step_momentum, 0, 1)
cand_b[6] = np.clip(cand_b[6] - step_micro, 0, 1) # Dim 6 sensitivity was negative-leaning

print(f"Current Best: {np.max(Y_all):.6f} at Index {best_idx}")
print(f"Cand A (Momentum): {'-'.join(f'{v:.6f}' for v in cand_a)}")
print(f"Cand B (Pivot):    {'-'.join(f'{v:.6f}' for v in cand_b)}")

Current Best: 9.899641 at Index 47
Cand A (Momentum): 0.102719-0.102828-0.213350-0.000000-1.000000-0.512206-0.213228-0.015300
Cand B (Pivot):    0.102719-0.102828-0.213350-0.000000-1.000000-0.512206-0.210228-0.015300


## 26 March - v1 - Coordinate Descent Nudge

In [6]:
# Use the absolute best point found so far
X_all = np.array(X)
Y_all = np.array(Y)
best_idx = np.argmax(Y_all)
x_best = X_all[best_idx]

# Define high-sensitivity dimensions from our previous analysis
# Dim 3 & 4 are locked. Dim 0 and 6 were the most 'active'.
target_dims = [0, 6, 1, 7] 
step_size = 0.005 # Ultra-fine precision

print(f"Current Best Y: {np.max(Y_all)}")
print("--- Generating Micro-Nudge Candidates ---")

for d in target_dims:
    # Nudge Up
    x_up = x_best.copy()
    x_up[d] = np.clip(x_up[d] + step_size, 0, 1)
    
    # Nudge Down
    x_down = x_best.copy()
    x_down[d] = np.clip(x_down[d] - step_size, 0, 1)
    
    print(f"Dim {d} [+] : {'-'.join(f'{v:.6f}' for v in x_up)}")
    print(f"Dim {d} [-] : {'-'.join(f'{v:.6f}' for v in x_down)}")

Current Best Y: 9.89944513359
--- Generating Micro-Nudge Candidates ---
Dim 0 [+] : 0.092719-0.102828-0.213350-0.000000-1.000000-0.512206-0.213228-0.015300
Dim 0 [-] : 0.082719-0.102828-0.213350-0.000000-1.000000-0.512206-0.213228-0.015300
Dim 6 [+] : 0.087719-0.102828-0.213350-0.000000-1.000000-0.512206-0.218228-0.015300
Dim 6 [-] : 0.087719-0.102828-0.213350-0.000000-1.000000-0.512206-0.208228-0.015300
Dim 1 [+] : 0.087719-0.107828-0.213350-0.000000-1.000000-0.512206-0.213228-0.015300
Dim 1 [-] : 0.087719-0.097828-0.213350-0.000000-1.000000-0.512206-0.213228-0.015300
Dim 7 [+] : 0.087719-0.102828-0.213350-0.000000-1.000000-0.512206-0.213228-0.020300
Dim 7 [-] : 0.087719-0.102828-0.213350-0.000000-1.000000-0.512206-0.213228-0.010300


## New Code 19 March - v2

In [6]:
import numpy as np
from sklearn.ensemble import RandomForestRegressor
from scipy.optimize import minimize
from scipy.stats import norm

# 1. SETUP DATA
X_arr = np.array(X)
Y_arr = np.array(Y).flatten()
current_best_y = np.max(Y_arr)

# 2. TRAIN ENSEMBLE
# Increasing n_estimators for smoother probability estimates
model = RandomForestRegressor(n_estimators=250, min_samples_split=2, random_state=42)
model.fit(X_arr, Y_arr)

# 3. DEFINE EXPECTED IMPROVEMENT (EI)
def expected_improvement(x):
    x = x.reshape(1, -1)
    # Get distribution from the forest
    tree_preds = np.array([tree.predict(x)[0] for tree in model.estimators_])
    mu = np.mean(tree_preds)
    sigma = np.std(tree_preds)
    
    if sigma == 0:
        return 0.0
    
    # Z-score of the improvement
    z = (mu - current_best_y) / sigma
    ei = (mu - current_best_y) * norm.cdf(z) + sigma * norm.pdf(z)
    return -ei  # Minimize negative EI

# # 4. TARGETED LOCAL SEARCH
# best_ei = 0
# best_x = X_arr[np.argmax(Y_arr)] # Start at the best known point

# for _ in range(100):
#     # Mix of local starts (near best) and global starts
#     if _ < 50:
#         x0 = np.clip(X_arr[np.argmax(Y_arr)] + np.random.normal(0, 0.1, 8), 0, 1)
#     else:
#         x0 = np.random.uniform(0, 1, 8)
        
#     res = minimize(expected_improvement, x0, bounds=[(0, 1)]*8, method='L-BFGS-B')
#     if -res.fun > best_ei:
#         best_ei = -res.fun
#         best_x = res.x

# Updated Step 4: High-Intensity Local Search
# We focus 80% of our effort within 0.1 distance of the BEST point
best_idx = np.argmax(Y_arr)
x_anchor = X_arr[best_idx]

for _ in range(200): # Increase iterations for finer resolution
    if _ < 160:
        # Local "Nudge": Start very close to the 9.899 point
        x0 = np.clip(x_anchor + np.random.normal(0, 0.05, 8), 0, 1)
    else:
        # Global "Safety": Check if we missed anything big elsewhere
        x0 = np.random.uniform(0, 1, 8)
        
    res = minimize(expected_improvement, x0, bounds=[(0, 1)]*8, method='L-BFGS-B')
    if -res.fun > best_ei:
        best_ei = -res.fun
        best_x = res.x


# 5. PREDICTION AT SUGGESTED POINT
final_preds = np.array([tree.predict(best_x.reshape(1, -1))[0] for tree in model.estimators_])
print(f"Recommended Point: {['-'.join(f'{v:.6f}' for v in best_x)]}")
print(f"Current Best Y: {current_best_y:.4f}")
print(f"Predicted Mean at New Point: {np.mean(final_preds):.4f}")
print(f"Expected Improvement (EI): {best_ei:.4f}")


Recommended Point: ['0.177523-0.106972-0.139817-0.000000-1.000000-0.479984-0.294613-0.100080']
Current Best Y: 9.8994
Predicted Mean at New Point: 9.6035
Expected Improvement (EI): 0.1253


In [7]:
# The Sensitivity Check: Which Variable is the "Trig point"?
import pandas as pd

def get_rf_sensitivity(x_point, model, epsilon=0.01):
    sensitivities = []
    base_pred = model.predict(x_point.reshape(1, -1))[0]
    
    for i in range(8):
        # Nudge up
        x_plus = x_point.copy()
        x_plus[i] = np.clip(x_plus[i] + epsilon, 0, 1)
        pred_plus = model.predict(x_plus.reshape(1, -1))[0]
        
        # Nudge down
        x_minus = x_point.copy()
        x_minus[i] = np.clip(x_minus[i] - epsilon, 0, 1)
        pred_minus = model.predict(x_minus.reshape(1, -1))[0]
        
        # Calculate local gradient (absolute change)
        grad = (abs(pred_plus - base_pred) + abs(pred_minus - base_pred)) / (2 * epsilon)
        sensitivities.append(grad)
        
    return np.array(sensitivities)

# Run sensitivity on your Recommended Point
recommended_x = np.array([0.177523, 0.106972, 0.139817, 0.0, 1.0, 0.479984, 0.294613, 0.100080])
sens = get_rf_sensitivity(recommended_x, model)

# Display Results
df_sens = pd.DataFrame({'Dimension': range(8), 'Sensitivity': sens})
print("\n--- Local Sensitivity at Recommended Point ---")
print(df_sens.sort_values(by='Sensitivity', ascending=False).to_string(index=False))


--- Local Sensitivity at Recommended Point ---
 Dimension  Sensitivity
         3     2.087123
         4     1.010834
         6     0.174268
         0     0.147088
         1     0.062966
         7     0.042755
         2     0.000000
         5     0.000000


In [8]:
# Updated Python Code: High-Precision 3D Search
# 1. IDENTIFY THE ANCHORS (Values from your best-ever 9.899 point)
x_best = X_arr[np.argmax(Y_arr)]

# 2. DEFINE A RESTRICTED SEARCH (Only Dims 0, 3, 4, 6 are "Live")
def restricted_objective(params):
    # params will only have 4 values
    x_full = x_best.copy()
    x_full[0] = params[0] # Dim 0
    x_full[3] = params[1] # Dim 3
    x_full[4] = params[2] # Dim 4
    x_full[6] = params[3] # Dim 6
    
    # We use pure Mean prediction now (Exploitation mode)
    return -model.predict(x_full.reshape(1, -1))[0]

# 3. RUN THE PRECISION OPTIMIZER
res_fine = minimize(
    restricted_objective, 
    x0=[x_best[0], x_best[3], x_best[4], x_best[6]], 
    bounds=[(0, 1), (0, 0.1), (0.9, 1), (0, 1)], # Tight bounds on sensitive Dims
    method='L-BFGS-B'
)

# 4. RECONSTRUCT FULL VECTOR
final_x = x_best.copy()
final_x[0], final_x[3], final_x[4], final_x[6] = res_fine.x

print(f"Precision Recommended Point: {['-'.join(f'{v:.6f}' for v in final_x)]}")
print(f"Predicted Y at Precision Point: {-res_fine.fun:.4f}")

Precision Recommended Point: ['0.087719-0.102828-0.213350-0.000000-1.000000-0.512206-0.213228-0.015300']
Predicted Y at Precision Point: 9.7278


In [9]:
# Final Optimization Update: "Simulated Annealing"
# HIGH-RESOLUTION SHAKE-UP
x_best = np.array([0.087719, 0.102828, 0.213350, 0.000000, 1.000000, 0.512206, 0.213228, 0.015300])

# We create 3 variations with very small "nudges"
# Focus on the most sensitive dimensions: 0, 3, 4, 6
nudge_intensity = 0.02 

variation_1 = x_best.copy()
variation_1[0] += nudge_intensity # Nudge Dim 0 (Sensitivity 0.14)

variation_2 = x_best.copy()
variation_2[6] -= nudge_intensity # Nudge Dim 6 (Sensitivity 0.17)

variation_3 = x_best.copy()
variation_3[0] -= nudge_intensity/2
variation_3[6] += nudge_intensity/2

print("Recommended 'Shake-up' Points:")
print(f"V1: {'-'.join(f'{v:.6f}' for v in variation_1)}")
print(f"V2: {'-'.join(f'{v:.6f}' for v in variation_2)}")
print(f"V3: {'-'.join(f'{v:.6f}' for v in variation_3)}")

Recommended 'Shake-up' Points:
V1: 0.107719-0.102828-0.213350-0.000000-1.000000-0.512206-0.213228-0.015300
V2: 0.087719-0.102828-0.213350-0.000000-1.000000-0.512206-0.193228-0.015300
V3: 0.077719-0.102828-0.213350-0.000000-1.000000-0.512206-0.223228-0.015300


## New Code 19 March - v1

In [5]:
import numpy as np
from sklearn.ensemble import RandomForestRegressor
from scipy.optimize import minimize
from scipy.stats import norm

# 1. SETUP DATA
X_arr = np.array(X)
Y_arr = np.array(Y).flatten()
current_best_y = np.max(Y_arr)

# 2. TRAIN ENSEMBLE
# Increasing n_estimators for smoother probability estimates
model = RandomForestRegressor(n_estimators=250, min_samples_split=2, random_state=42)
model.fit(X_arr, Y_arr)

# 3. DEFINE EXPECTED IMPROVEMENT (EI)
def expected_improvement(x):
    x = x.reshape(1, -1)
    # Get distribution from the forest
    tree_preds = np.array([tree.predict(x)[0] for tree in model.estimators_])
    mu = np.mean(tree_preds)
    sigma = np.std(tree_preds)
    
    if sigma == 0:
        return 0.0
    
    # Z-score of the improvement
    z = (mu - current_best_y) / sigma
    ei = (mu - current_best_y) * norm.cdf(z) + sigma * norm.pdf(z)
    return -ei  # Minimize negative EI

# 4. TARGETED LOCAL SEARCH
best_ei = 0
best_x = X_arr[np.argmax(Y_arr)] # Start at the best known point

for _ in range(100):
    # Mix of local starts (near best) and global starts
    if _ < 50:
        x0 = np.clip(X_arr[np.argmax(Y_arr)] + np.random.normal(0, 0.1, 8), 0, 1)
    else:
        x0 = np.random.uniform(0, 1, 8)
        
    res = minimize(expected_improvement, x0, bounds=[(0, 1)]*8, method='L-BFGS-B')
    if -res.fun > best_ei:
        best_ei = -res.fun
        best_x = res.x

# 5. PREDICTION AT SUGGESTED POINT
final_preds = np.array([tree.predict(best_x.reshape(1, -1))[0] for tree in model.estimators_])
print(f"Recommended Point: {['-'.join(f'{v:.6f}' for v in best_x)]}")
print(f"Current Best Y: {current_best_y:.4f}")
print(f"Predicted Mean at New Point: {np.mean(final_preds):.4f}")
print(f"Expected Improvement (EI): {best_ei:.4f}")


Recommended Point: ['0.175576-0.155598-0.207199-0.000000-1.000000-0.456312-0.290079-0.014202']
Current Best Y: 9.8994
Predicted Mean at New Point: 9.5399
Expected Improvement (EI): 0.1133


## New Code 5 March

In [7]:
# 1. Setup Data
X_arr = np.array(X)
Y_arr = np.array(Y).flatten()

# 2. Train Random Forest (The "Sharp" Surrogate)
# RF is better at "step functions" than MLP
model = RandomForestRegressor(n_estimators=100, min_samples_split=2, random_state=42)
model.fit(X_arr, Y_arr)

# 3. Define Acquisition Function (UCB)
# This rewards predictions that are high, but also rewards high variance (uncertainty)
def ucb_acquisition(x, model, kappa=1.5):
    # Get predictions from all 100 trees to estimate uncertainty (sigma)
    preds = np.array([tree.predict(x.reshape(1, -1))[0] for tree in model.estimators_])
    mu = np.mean(preds)
    sigma = np.std(preds)
    
    # UCB score: The "optimistic" prediction
    return -(mu + kappa * sigma)

# 4. Multi-start Optimization
best_x = None
best_val = np.inf

print("Optimizing with UCB acquisition...")
for _ in range(50):
    x0 = np.random.uniform(0, 1, 8)
    res = minimize(ucb_acquisition, x0, args=(model,), bounds=[(0, 1)]*8, method='L-BFGS-B')
    
    if res.fun < best_val:
        best_val = res.fun
        best_x = res.x

print(f"Recommended Submission: {'-'.join(f'{v:.6f}' for v in best_x)}")

tree_predictions = np.array([tree.predict(best_x.reshape(1, -1))[0] for tree in model.estimators_])
predicted_y = np.mean(tree_predictions)
model_uncertainty = np.std(tree_predictions)
print(f"Predicted Y (Mean): {predicted_y:.4f}    Model Uncertainty: {model_uncertainty:.4f}")

Optimizing with UCB acquisition...
Recommended Submission: 0.082682-0.917557-0.413349-0.038587-0.180483-0.597658-0.091566-0.094504
Predicted Y (Mean): 9.2816    Model Uncertainty: 0.5424


## New Code 26 Feb
This is NN

In [6]:
import numpy as np
import pandas as pd
from sklearn.neural_network import MLPRegressor
from sklearn.preprocessing import StandardScaler
from scipy.optimize import minimize
from scipy.spatial.distance import cdist

# 1. DATA PREPARATION
# X and Y are assumed to be your lists/arrays as provided
X_arr = np.array(X)
Y_arr = np.array(Y).reshape(-1, 1)

# Scalers are essential for Neural Networks to treat all 8 dims equally
scaler_y = StandardScaler()
Y_scaled = scaler_y.fit_transform(Y_arr).ravel()

# 2. TRAIN THE NEURAL NETWORK
# We use 'tanh' for smooth gradients and 'lbfgs' for small-sample precision
nn = MLPRegressor(
    hidden_layer_sizes=(64, 32),
    activation='tanh',
    solver='lbfgs',
    alpha=0.5,           # High regularization to smooth the "9.19 cliff"
    max_iter=5000,
    random_state=42
)
nn.fit(X_arr, Y_scaled)

# 3. DEFINE ROBUST OBJECTIVE (The "Safety Rope" Search)
best_idx = np.argmax(Y_arr)
x_best_ever = X_arr[best_idx]

def robust_objective(x):
    x_input = x.reshape(1, -1)
    
    # NN Predicted Score
    pred_scaled = nn.predict(x_input)[0]
    
    # Distance Penalty (Prevents sliding into "fake" corner peaks)
    # We penalize the Euclidean distance from our best known point
    distance = cdist(x_input, x_best_ever.reshape(1, -1), metric='euclidean')[0][0]
    
    # LAMBDA (0.3) controls how much we trust the NN vs. staying near data
    # Increase to 1.0 to stay very close to 9.899, decrease to 0.1 to explore more.
    penalty_weight = 0.3 
    return -pred_scaled + (penalty_weight * distance)

# 4. RUN CONSTRAINED OPTIMIZATION
# We start at the 9.899 point and look for a local improvement
res = minimize(
    robust_objective, 
    x0=x_best_ever, 
    bounds=[(0, 1)] * 8, 
    method='L-BFGS-B'
)

# 5. POST-PROCESS RESULTS
final_x = res.x
final_y_pred_scaled = nn.predict(final_x.reshape(1, -1))
final_y_pred = scaler_y.inverse_transform(final_y_pred_scaled.reshape(-1, 1))[0][0]

# 6. OUTPUTS
print("--- ROBUST NN OPTIMIZATION ---")
print(f"Current Best Score in Data: {Y_arr[best_idx][0]:.4f}")
print(f"NN Predicted Y at New Point: {final_y_pred:.4f}")
print("\nSuggested Submission (X0-X7):")
print('-'.join(f'{v:.6f}' for v in final_x))

# 7. FEATURE IMPORTANCE (Sensitivity at this specific point)
def get_sensitivity(x):
    eps = 1e-5
    grads = []
    base = nn.predict(x.reshape(1, -1))
    for i in range(len(x)):
        x_p = x.copy()
        x_p[i] += eps
        grads.append(abs(nn.predict(x_p.reshape(1, -1)) - base) / eps)
    return np.array(grads).flatten()

sens = get_sensitivity(final_x)
importance_df = pd.DataFrame({'Dim': range(8), 'Sensitivity': sens})
print("\nLocal Sensitivity (Which variables are pushing the score up here?):")
print(importance_df.sort_values(by='Sensitivity', ascending=False).to_string(index=False))


--- ROBUST NN OPTIMIZATION ---
Current Best Score in Data: 9.8994
NN Predicted Y at New Point: 9.7844

Suggested Submission (X0-X7):
0.000000-0.000000-0.000000-0.000000-1.000000-0.508290-0.000000-0.079755

Local Sensitivity (Which variables are pushing the score up here?):
 Dim  Sensitivity
   2     0.378054
   0     0.274181
   6     0.234103
   1     0.111094
   3     0.102566
   7     0.057420
   4     0.012994
   5     0.003478


## New Code 18 Feb

In [10]:
# 1. SETUP
# (Assuming X and Y are updated with your new 9.899 point)
n_dims = X.shape[1]
y_max = np.max(Y)

# 1. FIXING THE NOISE AND LENGTH SCALE BOUNDS
kernel = C(1.0, (1e-2, 1e4)) * Matern(
    length_scale=[1.0] * n_dims, 
    length_scale_bounds=(1e-2, 1000.0), # Increased to 1000 to clear warning
    nu=2.5
) + WhiteKernel(
    noise_level=1e-8, 
    noise_level_bounds=(1e-15, 1e-3) # Lowered to clear 1e-05 warning
)

model = GaussianProcessRegressor(
    kernel=kernel,
    alpha=1e-10, # Very small alpha since we want to trust our data points
    normalize_y=True,
    n_restarts_optimizer=50
)
model.fit(X, Y)
print(f"Optimized Kernel: {model.kernel_}")

Optimized Kernel: 6.09**2 * Matern(length_scale=[4.36, 6.34, 3.38, 7.49, 19.7, 6.28, 4.43, 1e+03], nu=2.5) + WhiteKernel(noise_level=6.92e-14)


C:\Users\Dave Amin\AppData\Roaming\Python\Python313\site-packages\sklearn\gaussian_process\kernels.py:450: ConvergenceWarning: The optimal value found for dimension 7 of parameter k1__k2__length_scale is close to the specified upper bound 1000.0. Increasing the bound and calling fit again may find a better value.
  warnings.warn(


In [12]:
# 2. DUAL-REGION TARGETED SAMPLING
# Best from "Old Region"
x_old_best = np.array([0.087, 0.102, 0.213, 0.0, 1.0, 0.512, 0.213, 0.015])
# Best from "New Region" (Your latest output)
x_new_best = np.array([0.0, 0.0, 0.0, 0.754, 1.0, 0.845, 0.0, 0.0])

num_samples = 100000
# Jitter both regions
samples_old = np.random.normal(loc=x_old_best, scale=0.1, size=(num_samples, n_dims))
samples_new = np.random.normal(loc=x_new_best, scale=0.1, size=(num_samples, n_dims))

x_candidates = np.concatenate([
    samples_old, 
    samples_new, 
    np.random.uniform(0, 1, size=(50000, n_dims)),
    X[np.argsort(Y)[-5:]]
], axis=0)
x_candidates = np.clip(x_candidates, 0, 1)

# 3. USE A DYNAMIC BETA
# We check a high-exploration (beta=3.0) and a high-exploitation (beta=1.0)
ucb_high = get_ucb(x_candidates, model, beta=3.0)
ucb_low = get_ucb(x_candidates, model, beta=1.0)

starts = np.concatenate([
    x_candidates[np.argsort(ucb_high)[-5:]],
    x_candidates[np.argsort(ucb_low)[-5:]]
])

# 5. THE POLISHER
best_res_x = None
best_val = -np.inf

for x0 in starts:
    res = minimize(lambda x: -get_ucb(x, model), x0=x0, bounds=[(0, 1)] * n_dims, method='L-BFGS-B')
    if -res.fun > best_val:
        best_val = -res.fun
        best_res_x = res.x
x_final = np.clip(best_res_x, 0, 1)

In [13]:
# 6. RESULTS
print(f"Weekly Submission: {[f'{v:.8f}' for v in x_final]}")
print(f"Predicted Y (Upper Bound): {best_val:.6f}")
formatted_str = "-".join(f"{v:.6f}" for v in x_final)
print(f"Submit this: {formatted_str}")

Weekly Submission: ['0.00000000', '0.00000000', '0.00000000', '0.84544978', '1.00000000', '0.81116054', '0.00000000', '0.00000000']
Predicted Y (Upper Bound): 10.188169
Submit this: 0.000000-0.000000-0.000000-0.845450-1.000000-0.811161-0.000000-0.000000


In [ ]:
results
1. first new code
Predicted Y (Upper Bound): 10.175946
Submit this: 0.000000-0.000000-0.000000-0.754065-1.000000-0.845992-0.000000-0.000000
2. second new code
Predicted Y (Upper Bound): 10.188169
Submit this: 0.000000-0.000000-0.000000-0.845450-1.000000-0.811161-0.000000-0.000000

## New Code 8 Feb - v2.3 - final jitter

In [18]:
# 1. SETUP THE JITTER SEARCH
# Using your latest result as the anchor
x_anchor = np.array([0.08062398, 0.05075951, 0.24626210, 0.00000000, 
                     1.00000000, 0.50312490, 0.18516555, 0.00000000])

# 2. GENERATE TIGHT CLOUD
# We generate points very close to the anchor (std dev of 0.02)
num_jitter = 300000
x_jitter = np.random.normal(loc=x_anchor, scale=0.02, size=(num_jitter, n_dims))
x_jitter = np.clip(x_jitter, 0, 1)

# Include the anchor itself
x_candidates = np.vstack([x_jitter, x_anchor])

# 3. PURE EXPLOITATION (Mean Prediction Only)
# At this point, we care only about the mean, not the uncertainty (Beta = 0)
mu_vals = model.predict(x_candidates)

# 4. PICK THE ABSOLUTE BEST
best_idx = np.argmax(mu_vals)
x_final_jitter = x_candidates[best_idx]

# 5. FINAL OUTPUT
print(f"Weekly Submission: {[f'{v:.8f}' for v in x_final_jitter]}")
print(f"Submit this: {'-'.join(f'{v:.6f}' for v in x_final_jitter)}")
print(f"Predicted Y: {mu_vals[best_idx]:.6f}")

Weekly Submission: ['0.08771926', '0.10282812', '0.21335020', '0.00000000', '1.00000000', '0.51220570', '0.21322786', '0.01530003']
Submit this: 0.087719-0.102828-0.213350-0.000000-1.000000-0.512206-0.213228-0.015300
Predicted Y: 10.022720


## New Code 8 Feb - v2.2 - add a Sobol Sequence

In [17]:
# ... (Keep previous imports and kernel setup)

# 1. SOBOL SEQUENCE GENERATION (High-coverage sampling)
sampler = qmc.Sobol(d=n_dims, scramble=True)
sample_power = 18 # 2^18 = 262,144 points
x_sobol = sampler.random_base2(m=sample_power)

# 2. TARGETED LOCAL REFINEMENT
# Focus specifically on the "unsettled" dimensions (2 and 6) 
# while keeping others near their discovered optimal boundaries
x_best_observed = X[np.argmax(Y)]
x_refined = np.tile(x_final, (50000, 1)) 
# Add small noise only to the variables that are still shifting
x_refined[:, [2, 6]] += np.random.normal(0, 0.05, size=(50000, 2))
x_refined = np.clip(x_refined, 0, 1)

x_candidates = np.concatenate([x_sobol, x_refined, X[np.argsort(Y)[-5:]]], axis=0)

# 3. EXPLOITATION-HEAVY ACQUISITION
# We use a very low Beta (1.0) to focus on the peak of the mountain
def get_exploitation_ucb(x, model, beta=1.0):
    mu, sigma = model.predict(np.atleast_2d(x), return_std=True)
    return mu + beta * sigma

scores = get_exploitation_ucb(x_candidates, model)
starts = x_candidates[np.argsort(scores)[-8:]]

# 4. FINAL POLISH
best_res_x = None
best_val = -np.inf

for x0 in starts:
    res = minimize(lambda x: -get_exploitation_ucb(x, model), 
                   x0=x0, bounds=[(0, 1)] * n_dims, method='L-BFGS-B')
    if -res.fun > best_val:
        best_val = -res.fun
        best_res_x = res.x

x_final_refined = np.clip(best_res_x, 0, 1)

# 5. FINAL OUTPUT
print(f"Weekly Submission: {[f'{v:.8f}' for v in x_final_refined]}")
print(f"Submit this: {'-'.join(f'{v:.6f}' for v in x_final_refined)}")
print(f"Predicted Y: {model.predict(x_final_refined.reshape(1, -1))[0]:.6f}")

Weekly Submission: ['0.08062398', '0.05075951', '0.24626210', '0.00000000', '1.00000000', '0.50312490', '0.18516555', '0.00000000']
Submit this: 0.080624-0.050760-0.246262-0.000000-1.000000-0.503125-0.185166-0.000000
Predicted Y: 10.015190


## New Code 8 Feb - v2.1 - Localised Search

In [15]:
# 1. DEFINE THE LOCAL AREA
# Best point from your Y data (index 41)
x_best_observed = X[np.argmax(Y)] 

# 2. HYBRID CANDIDATE GENERATION
num_global = 150000
num_local = 150000

# Global candidates
x_global = np.random.uniform(0, 1, size=(num_global, n_dims))

# Local candidates: Search within a 0.15 radius of the best observed point
x_local = np.random.normal(loc=x_best_observed, scale=0.15, size=(num_local, n_dims))
x_local = np.clip(x_local, 0, 1) # Keep within bounds

x_candidates = np.concatenate([x_global, x_local, X[np.argsort(Y)[-5:]]], axis=0)

# 3. EVALUATE ACQUISITION
# We'll use a slightly lower Beta for UCB to stay grounded
def get_ucb(x, model, beta=2.0):
    mu, sigma = model.predict(np.atleast_2d(x), return_std=True)
    return mu + beta * sigma

ei_vals = get_ei(x_candidates, model, y_max)
ucb_vals = get_ucb(x_candidates, model)

# Pick the top 4 EI and top 4 UCB candidates as starting points
starts = np.concatenate([
    x_candidates[np.argsort(ei_vals)[-4:]],
    x_candidates[np.argsort(ucb_vals)[-4:]]
])

# 4. POLISH
best_res_x = None
best_val = -np.inf

for x0 in starts:
    # We maximize UCB for the final recommendation
    res = minimize(lambda x: -get_ucb(x, model), x0=x0, bounds=[(0, 1)] * n_dims, method='L-BFGS-B')
    if -res.fun > best_val:
        best_val = -res.fun
        best_res_x = res.x

x_final = np.clip(best_res_x, 0, 1)

# 5. FINAL OUTPUT
print(f"Weekly Submission: {[f'{v:.8f}' for v in x_final]}")
print(f"Submit this: {'-'.join(f'{v:.6f}' for v in x_final)}")
print(f"Predicted Y: {model.predict(x_final.reshape(1, -1))[0]:.6f}")

Weekly Submission: ['0.00000000', '0.00000000', '0.28992028', '0.00000000', '1.00000000', '0.54253155', '0.08216634', '0.00000000']
Submit this: 0.000000-0.000000-0.289920-0.000000-1.000000-0.542532-0.082166-0.000000
Predicted Y: 9.956926


## New Code 8 Feb - v2

In [12]:
# 1. UPDATED KERNEL BOUNDS
# Increased length_scale upper bound to 500 to stop the convergence warning
kernel = C(1.0, (1e-2, 1e4)) * Matern(
    length_scale=[0.5] * n_dims, 
    length_scale_bounds=(1e-2, 500.0), 
    nu=2.5
) + WhiteKernel(
    noise_level=1e-8, 
    noise_level_bounds=(1e-15, 1e-3)  )
model = GaussianProcessRegressor(
    kernel=kernel,
    alpha=1e-8,
    normalize_y=True,
    n_restarts_optimizer=50 # Increased for better global fit
)
model.fit(X, Y)
print(f"Optimized Kernel: {model.kernel_}")

Optimized Kernel: 6.62**2 * Matern(length_scale=[4.55, 6.43, 3.39, 8.67, 23.1, 6.14, 4.56, 500], nu=2.5) + WhiteKernel(noise_level=1e-15)


C:\ProgramData\anaconda3\Lib\site-packages\sklearn\gaussian_process\kernels.py:450: ConvergenceWarning: The optimal value found for dimension 7 of parameter k1__k2__length_scale is close to the specified upper bound 500.0. Increasing the bound and calling fit again may find a better value.
  warnings.warn(
C:\ProgramData\anaconda3\Lib\site-packages\sklearn\gaussian_process\kernels.py:440: ConvergenceWarning: The optimal value found for dimension 0 of parameter k2__noise_level is close to the specified lower bound 1e-15. Decreasing the bound and calling fit again may find a better value.
  warnings.warn(


In [13]:
# 2. HYBRID ACQUISITION (UCB + EI)
def get_ucb(x_points, model, beta=2.5):
    x_points = np.atleast_2d(x_points)
    mu, sigma = model.predict(x_points, return_std=True)
    return mu + beta * sigma

In [14]:
# 3. MULTI-STRATEGY SEARCH
# We will look for the best point using both EI and UCB
num_candidates = 250000
x_candidates = np.random.uniform(0, 1, size=(num_candidates, n_dims))
x_candidates = np.concatenate([x_candidates, X[np.argsort(Y)[-5:]]], axis=0)

# Evaluate both metrics
ei_vals = get_ei(x_candidates, model, y_max)
ucb_vals = get_ucb(x_candidates, model)

# Take top starts from both to ensure we don't miss a peak
starts = np.concatenate([
    x_candidates[np.argsort(ei_vals)[-3:]],
    x_candidates[np.argsort(ucb_vals)[-3:]]
])

best_res_x = None
best_val = -np.inf

# Run the polisher on the hybrid starts
for x0 in starts:
    # Optimizing for UCB here to find the highest potential peak
    res = minimize(lambda x: -get_ucb(x, model), x0=x0, bounds=[(0, 1)] * n_dims, method='L-BFGS-B')
    if -res.fun > best_val:
        best_val = -res.fun
        best_res_x = res.x

x_final = np.clip(best_res_x, 0, 1)

# 4. FINAL OUTPUT
print(f"Weekly Submission: {[f'{v:.8f}' for v in x_final]}")
print(f"Submit this: {'-'.join(f'{v:.6f}' for v in x_final)}")
print(f"Predicted Y (Mean): {model.predict(x_final.reshape(1, -1))[0]:.6f}")

Weekly Submission: ['0.00000000', '0.00000000', '0.31218099', '0.00000000', '1.00000000', '0.56563217', '0.00000000', '0.00000000']
Submit this: 0.000000-0.000000-0.312181-0.000000-1.000000-0.565632-0.000000-0.000000
Predicted Y (Mean): 9.903926


## New Code 8 Feb - v1

In [9]:
# 1. SETUP & DATA
X = np.array(X) # Loaded from your data
Y = np.array(Y) # Loaded from your data
n_dims = X.shape[1]
y_max = np.max(Y)

# 2. IMPROVED KERNEL
# Increased n_restarts and adjusted length_scale bounds for [0, 1] space
kernel = C(1.0, (1e-2, 1e3)) * Matern(
    length_scale=[0.5] * n_dims, 
    length_scale_bounds=(1e-2, 1e2), # 100 is plenty for a 0-1 range
    nu=2.5
) + WhiteKernel(
    noise_level=1e-5, 
    noise_level_bounds=(1e-12, 1e-2)   )

model = GaussianProcessRegressor(
    kernel=kernel,
    alpha=1e-6, # Slightly higher alpha for numerical stability
    normalize_y=True,
    n_restarts_optimizer=25 # Sufficient restarts for 8D
)
print("Fitting model...")
model.fit(X, Y)
print(f"Model Kernel: {model.kernel_}")

Fitting model...
Model Kernel: 5.81**2 * Matern(length_scale=[4.24, 5.83, 3.07, 9.05, 22.5, 5.56, 4.26, 100], nu=2.5) + WhiteKernel(noise_level=2.92e-12)


C:\ProgramData\anaconda3\Lib\site-packages\sklearn\gaussian_process\kernels.py:450: ConvergenceWarning: The optimal value found for dimension 7 of parameter k1__k2__length_scale is close to the specified upper bound 100.0. Increasing the bound and calling fit again may find a better value.
  warnings.warn(


In [10]:
# 3. VECTORIZED EI FUNCTION
def get_ei(x_points, model, y_max):
    x_points = np.atleast_2d(x_points)
    mu, sigma = model.predict(x_points, return_std=True)
    sigma = sigma.reshape(-1, 1)
    mu = mu.reshape(-1, 1)
    with np.errstate(divide='warn'):
        imp = mu - y_max
        Z = imp / sigma
        ei = imp * norm.cdf(Z) + sigma * norm.pdf(Z)
        ei[sigma <= 0.0] = 0.0
    return ei.flatten()

In [11]:
# 4. CANDIDATE GENERATION (Large Scale)
num_candidates = 250000 
x_candidates = np.random.uniform(0, 1, size=(num_candidates, n_dims))
# Include current best points and their slight perturbations
best_indices = np.argsort(Y)[-5:]
x_candidates = np.concatenate([x_candidates, X[best_indices]], axis=0)

ei_vals = get_ei(x_candidates, model, y_max)
# Get top 5 starting points for multi-start optimization
top_idx = np.argsort(ei_vals)[-5:]
starts = x_candidates[top_idx]

# 5. MULTI-START POLISHER
best_res_x = None
best_res_ei = -np.inf

def objective(x):
    return -get_ei(x, model, y_max)[0]

for x0 in starts:
    res = minimize(objective, x0=x0, bounds=[(0, 1)] * n_dims, method='L-BFGS-B')
    if -res.fun > best_res_ei:
        best_res_ei = -res.fun
        best_res_x = res.x

x_final = np.clip(best_res_x, 0, 1)

# 6. RESULTS
print(f"Weekly Submission: {[f'{v:.8f}' for v in x_final]}")
print(f"Submit this: {'-'.join(f'{v:.6f}' for v in x_final)}")
print(f"Predicted Y: {model.predict(x_final.reshape(1, -1))[0]:.6f}")

Weekly Submission: ['0.09231048', '0.09794269', '0.22822354', '0.00000000', '1.00000000', '0.53313993', '0.20697211', '0.00000000']
Submit this: 0.092310-0.097943-0.228224-0.000000-1.000000-0.533140-0.206972-0.000000
Predicted Y: 10.071498


## Results on 8 Feb:


Weekly Submission: ['0.09231048', '0.09794269', '0.22822354', '0.00000000', '1.00000000', '0.53313993', '0.20697211', '0.00000000']
Submit this: 0.092310-0.097943-0.228224-0.000000-1.000000-0.533140-0.206972-0.000000
Predicted Y: 10.071498

v2
Weekly Submission: ['0.00000000', '0.00000000', '0.31218099', '0.00000000', '1.00000000', '0.56563217', '0.00000000', '0.00000000']
Submit this: 0.000000-0.000000-0.312181-0.000000-1.000000-0.565632-0.000000-0.000000
Predicted Y (Mean): 9.903926

v2.1
Weekly Submission: ['0.00000000', '0.00000000', '0.28992028', '0.00000000', '1.00000000', '0.54253155', '0.08216634', '0.00000000']
Submit this: 0.000000-0.000000-0.289920-0.000000-1.000000-0.542532-0.082166-0.000000
Predicted Y: 9.956926

v2.2 with sobol
Weekly Submission: ['0.08062398', '0.05075951', '0.24626210', '0.00000000', '1.00000000', '0.50312490', '0.18516555', '0.00000000']
Submit this: 0.080624-0.050760-0.246262-0.000000-1.000000-0.503125-0.185166-0.000000
Predicted Y: 10.015190

v2.3 with final jitter
Weekly Submission: ['0.08771926', '0.10282812', '0.21335020', '0.00000000', '1.00000000', '0.51220570', '0.21322786', '0.01530003']
Submit this: 0.087719-0.102828-0.213350-0.000000-1.000000-0.512206-0.213228-0.015300
Predicted Y: 10.022720


## Misc

In [14]:
print("Max Y =", max(Y))

Max Y = 9.89944513359
